// @clerk/backend-core/dist/cjs/types/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenVerificationErrorReason = exports.AuthErrorReason = void 0;
var AuthErrorReason;
(function (AuthErrorReason) {
    AuthErrorReason["InlineKeyMissing"] = "inline-key-missing";
    AuthErrorReason["InlineKeyInvalid"] = "inline-key-invalid";
    AuthErrorReason["HeaderExpired"] = "header-expired";
    AuthErrorReason["HeaderEarly"] = "header-early";
    AuthErrorReason["HeaderInvalid"] = "header-invalid";
    AuthErrorReason["HeaderInvalidIssuer"] = "header-invalid-issuer";
    AuthErrorReason["HeaderMissingNonBrowser"] = "header-missing-non-browser";
    AuthErrorReason["HeaderMissingCORS"] = "header-missing-cors";
    AuthErrorReason["HeaderUnauthorizedParty"] = "header-unauthorized-party";
    AuthErrorReason["HeaderVerificationFailed"] = "header-verification-failed";
    AuthErrorReason["UATMissing"] = "uat-missing";
    AuthErrorReason["CrossOriginReferrer"] = "cross-origin-referrer";
    AuthErrorReason["CookieAndUATMissing"] = "cookie-and-uat-missing";
    AuthErrorReason["StandardOut"] = "standard-out";
    AuthErrorReason["CookieMissing"] = "cookie-missing";
    AuthErrorReason["CookieExpired"] = "cookie-expired";
    AuthErrorReason["CookieEarly"] = "cookie-early";
    AuthErrorReason["CookieInvalid"] = "cookie-invalid";
    AuthErrorReason["CookieInvalidIssuer"] = "cookie-invalid-issuer";
    AuthErrorReason["CookieOutDated"] = "cookie-outdated";
    AuthErrorReason["CookieUnauthorizedParty"] = "cookie-unauthorized-party";
    AuthErrorReason["CookieVerificationFailed"] = "cookie-verification-failed";
    AuthErrorReason["InternalError"] = "internal-error";
    AuthErrorReason["PublicKeyFetchError"] = "pk-fetch-error";
    AuthErrorReason["Unknown"] = "unknown";
})(AuthErrorReason = exports.AuthErrorReason || (exports.AuthErrorReason = {}));
var TokenVerificationErrorReason;
(function (TokenVerificationErrorReason) {
    TokenVerificationErrorReason[TokenVerificationErrorReason["Expired"] = 0] = "Expired";
    TokenVerificationErrorReason[TokenVerificationErrorReason["Invalid"] = 1] = "Invalid";
    TokenVerificationErrorReason[TokenVerificationErrorReason["ImportKeyError"] = 2] = "ImportKeyError";
    TokenVerificationErrorReason[TokenVerificationErrorReason["JWTKeyMissing"] = 3] = "JWTKeyMissing";
    TokenVerificationErrorReason[TokenVerificationErrorReason["MalformedToken"] = 4] = "MalformedToken";
    TokenVerificationErrorReason[TokenVerificationErrorReason["VerificationFailed"] = 5] = "VerificationFailed";
    TokenVerificationErrorReason[TokenVerificationErrorReason["NotActiveYet"] = 6] = "NotActiveYet";
    TokenVerificationErrorReason[TokenVerificationErrorReason["InvalidIssuer"] = 7] = "InvalidIssuer";
    TokenVerificationErrorReason[TokenVerificationErrorReason["UnauthorizedParty"] = 8] = "UnauthorizedParty";
    TokenVerificationErrorReason[TokenVerificationErrorReason["PublicKeyFetchError"] = 9] = "PublicKeyFetchError";
})(TokenVerificationErrorReason = exports.TokenVerificationErrorReason || (exports.TokenVerificationErrorReason = {}));
