// @clerk/edge/dist/cjs/vercel-edge/ClerkAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClerkAPI = void 0;
const backend_core_1 = require("@clerk/backend-core");
const info_1 = require("../info");
exports.ClerkAPI = new backend_core_1.ClerkBackendAPI({
    libName: info_1.LIB_NAME,
    libVersion: info_1.LIB_VERSION,
    apiClient: {
        async request({ url, method, queryParams, headerParams, bodyParams }) {
            const finalUrl = new URL(url || '');
            if (queryParams) {
                for (const [key, val] of Object.entries(queryParams)) {
                    if (val) {
                        [val].flat().forEach(v => finalUrl.searchParams.append(key, v));
                    }
                }
            }
            const response = await fetch(finalUrl.href, {
                method,
                headers: {
                    ...headerParams,
                    'X-Clerk-SDK': `vercel-edge/${info_1.LIB_VERSION}`,
                },
                ...(bodyParams && Object.keys(bodyParams).length > 0 && { body: JSON.stringify(bodyParams) }),
            });
            const isJSONResponse = headerParams && headerParams['Content-Type'] === 'application/json';
            const data = await (isJSONResponse ? response.json() : response.text());
            if (!response.ok) {
                throw new backend_core_1.ClerkAPIResponseError(response.statusText, {
                    data: (data === null || data === void 0 ? void 0 : data.errors) || data,
                    status: response.status,
                });
            }
            return data;
        },
    },
});
