// @clerk/edge/dist/mjs/vercel-edge/ClerkAPI.js
import { ClerkAPIResponseError, ClerkBackendAPI } from '@clerk/backend-core';
import { LIB_NAME, LIB_VERSION } from '../info';
export const ClerkAPI = new ClerkBackendAPI({
    libName: LIB_NAME,
    libVersion: LIB_VERSION,
    apiClient: {
        async request({ url, method, queryParams, headerParams, bodyParams }) {
            const finalUrl = new URL(url || '');
            if (queryParams) {
                for (const [key, val] of Object.entries(queryParams)) {
                    if (val) {
                        [val].flat().forEach(v => finalUrl.searchParams.append(key, v));
                    }
                }
            }
            const response = await fetch(finalUrl.href, {
                method,
                headers: {
                    ...headerParams,
                    'X-Clerk-SDK': `vercel-edge/${LIB_VERSION}`,
                },
                ...(bodyParams && Object.keys(bodyParams).length > 0 && { body: JSON.stringify(bodyParams) }),
            });
            const isJSONResponse = headerParams && headerParams['Content-Type'] === 'application/json';
            const data = await (isJSONResponse ? response.json() : response.text());
            if (!response.ok) {
                throw new ClerkAPIResponseError(response.statusText, {
                    data: data?.errors || data,
                    status: response.status,
                });
            }
            return data;
        },
    },
});
