// @clerk/backend-core/dist/mjs/api/endpoints/OrganizationApi.js
import { joinPaths } from '../../util/path';
import { AbstractAPI } from './AbstractApi';
const basePath = '/organizations';
export class OrganizationAPI extends AbstractAPI {
    async getOrganizationList(params) {
        return this.APIClient.request({
            method: 'GET',
            path: basePath,
            queryParams: params,
        });
    }
    async createOrganization(params) {
        return this.APIClient.request({
            method: 'POST',
            path: basePath,
            bodyParams: params,
        });
    }
    async getOrganization(params) {
        const organizationIdOrSlug = 'organizationId' in params ? params.organizationId : params.slug;
        this.requireId(organizationIdOrSlug);
        return this.APIClient.request({
            method: 'GET',
            path: joinPaths(basePath, organizationIdOrSlug),
        });
    }
    async updateOrganization(organizationId, params) {
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'PATCH',
            path: joinPaths(basePath, organizationId),
            bodyParams: params,
        });
    }
    async updateOrganizationMetadata(organizationId, params) {
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'PATCH',
            path: joinPaths(basePath, organizationId, 'metadata'),
            bodyParams: params,
        });
    }
    async deleteOrganization(organizationId) {
        return this.APIClient.request({
            method: 'DELETE',
            path: joinPaths(basePath, organizationId),
        });
    }
    async getOrganizationMembershipList(params) {
        const { organizationId, limit, offset } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'GET',
            path: joinPaths(basePath, organizationId, 'memberships'),
            queryParams: { limit, offset },
        });
    }
    async createOrganizationMembership(params) {
        const { organizationId, userId, role } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'POST',
            path: joinPaths(basePath, organizationId, 'memberships'),
            bodyParams: {
                userId,
                role,
            },
        });
    }
    async updateOrganizationMembership(params) {
        const { organizationId, userId, role } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'PATCH',
            path: joinPaths(basePath, organizationId, 'memberships', userId),
            bodyParams: {
                role,
            },
        });
    }
    async updateOrganizationMembershipMetadata(params) {
        const { organizationId, userId, publicMetadata, privateMetadata } = params;
        return this.APIClient.request({
            method: 'PATCH',
            path: joinPaths(basePath, organizationId, 'memberships', userId, 'metadata'),
            bodyParams: {
                publicMetadata,
                privateMetadata,
            },
        });
    }
    async deleteOrganizationMembership(params) {
        const { organizationId, userId } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'DELETE',
            path: joinPaths(basePath, organizationId, 'memberships', userId),
        });
    }
    async getPendingOrganizationInvitationList(params) {
        const { organizationId, limit, offset } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'GET',
            path: joinPaths(basePath, organizationId, 'invitations', 'pending'),
            queryParams: { limit, offset },
        });
    }
    async createOrganizationInvitation(params) {
        const { organizationId, ...bodyParams } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'POST',
            path: joinPaths(basePath, organizationId, 'invitations'),
            bodyParams: { ...bodyParams },
        });
    }
    async revokeOrganizationInvitation(params) {
        const { organizationId, invitationId, requestingUserId } = params;
        this.requireId(organizationId);
        return this.APIClient.request({
            method: 'POST',
            path: joinPaths(basePath, organizationId, 'invitations', invitationId, 'revoke'),
            bodyParams: {
                requestingUserId,
            },
        });
    }
}
