// @clerk/backend-core/dist/cjs/api/resources/Client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const Session_1 = require("./Session");
class Client {
    constructor(id, sessionIds, sessions, signInAttemptId, signUpAttemptId, signInId, signUpId, lastActiveSessionId, createdAt, updatedAt) {
        this.id = id;
        this.sessionIds = sessionIds;
        this.sessions = sessions;
        this.signInAttemptId = signInAttemptId;
        this.signUpAttemptId = signUpAttemptId;
        this.signInId = signInId;
        this.signUpId = signUpId;
        this.lastActiveSessionId = lastActiveSessionId;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }
    static fromJSON(data) {
        return new Client(data.id, data.session_ids, data.sessions.map(x => Session_1.Session.fromJSON(x)), data.sign_in_attempt_id, data.sign_up_attempt_id, data.sign_in_id, data.sign_up_id, data.last_active_session_id, data.created_at, data.updated_at);
    }
}
exports.Client = Client;
