// @clerk/backend-core/dist/cjs/api/resources/ExternalAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalAccount = void 0;
const Verification_1 = require("./Verification");
class ExternalAccount {
    constructor(id, provider, identificationId, externalId, approvedScopes, emailAddress, firstName, lastName, picture, username, publicMetadata = {}, label, verification) {
        this.id = id;
        this.provider = provider;
        this.identificationId = identificationId;
        this.externalId = externalId;
        this.approvedScopes = approvedScopes;
        this.emailAddress = emailAddress;
        this.firstName = firstName;
        this.lastName = lastName;
        this.picture = picture;
        this.username = username;
        this.publicMetadata = publicMetadata;
        this.label = label;
        this.verification = verification;
    }
    static fromJSON(data) {
        return new ExternalAccount(data.id, data.provider, data.identification_id, data.provider_user_id, data.approved_scopes, data.email_address, data.first_name, data.last_name, data.avatar_url, data.username, data.public_metadata, data.label, data.verification && Verification_1.Verification.fromJSON(data.verification));
    }
}
exports.ExternalAccount = ExternalAccount;
