// @clerk/backend-core/dist/cjs/api/resources/JSON.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectType = void 0;
var ObjectType;
(function (ObjectType) {
    ObjectType["AllowlistIdentifier"] = "allowlist_identifier";
    ObjectType["Client"] = "client";
    ObjectType["Email"] = "email";
    ObjectType["EmailAddress"] = "email_address";
    ObjectType["ExternalAccount"] = "external_account";
    ObjectType["FacebookAccount"] = "facebook_account";
    ObjectType["GoogleAccount"] = "google_account";
    ObjectType["Invitation"] = "invitation";
    ObjectType["Organization"] = "organization";
    ObjectType["OrganizationInvitation"] = "organization_invitation";
    ObjectType["OrganizationMembership"] = "organization_membership";
    ObjectType["PhoneNumber"] = "phone_number";
    ObjectType["RedirectUrl"] = "redirect_url";
    ObjectType["Session"] = "session";
    ObjectType["SignInAttempt"] = "sign_in_attempt";
    ObjectType["SignInToken"] = "sign_in_token";
    ObjectType["SignUpAttempt"] = "sign_up_attempt";
    ObjectType["SmsMessage"] = "sms_message";
    ObjectType["User"] = "user";
    ObjectType["Web3Wallet"] = "web3_wallet";
    ObjectType["Token"] = "token";
    ObjectType["TotalCount"] = "total_count";
})(ObjectType = exports.ObjectType || (exports.ObjectType = {}));
