// @clerk/backend-core/dist/cjs/api/resources/OrganizationMembership.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationMembershipPublicUserData = exports.OrganizationMembership = void 0;
const resources_1 = require("../resources");
class OrganizationMembership {
    constructor(id, role, publicMetadata = {}, privateMetadata = {}, createdAt, updatedAt, organization, publicUserData) {
        this.id = id;
        this.role = role;
        this.publicMetadata = publicMetadata;
        this.privateMetadata = privateMetadata;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.organization = organization;
        this.publicUserData = publicUserData;
    }
    static fromJSON(data) {
        return new OrganizationMembership(data.id, data.role, data.public_metadata, data.private_metadata, data.created_at, data.updated_at, resources_1.Organization.fromJSON(data.organization), OrganizationMembershipPublicUserData.fromJSON(data.public_user_data));
    }
}
exports.OrganizationMembership = OrganizationMembership;
class OrganizationMembershipPublicUserData {
    constructor(identifier, firstName, lastName, profileImageUrl, userId) {
        this.identifier = identifier;
        this.firstName = firstName;
        this.lastName = lastName;
        this.profileImageUrl = profileImageUrl;
        this.userId = userId;
    }
    static fromJSON(data) {
        return new OrganizationMembershipPublicUserData(data.identifier, data.first_name, data.last_name, data.profile_image_url, data.user_id);
    }
}
exports.OrganizationMembershipPublicUserData = OrganizationMembershipPublicUserData;
