// @clerk/backend-core/dist/cjs/api/errors/Errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClerkAPIResponseError = exports.parseError = exports.parseErrors = void 0;
function parseErrors(data = []) {
    return data.length > 0 ? data.map(parseError) : [];
}
exports.parseErrors = parseErrors;
function parseError(error) {
    var _a, _b;
    return {
        code: error.code,
        message: error.message,
        longMessage: error.long_message,
        meta: {
            paramName: (_a = error === null || error === void 0 ? void 0 : error.meta) === null || _a === void 0 ? void 0 : _a.param_name,
            sessionId: (_b = error === null || error === void 0 ? void 0 : error.meta) === null || _b === void 0 ? void 0 : _b.session_id,
        },
    };
}
exports.parseError = parseError;
class ClerkAPIResponseError extends Error {
    constructor(message, { data, status }) {
        super(message);
        Object.setPrototypeOf(this, ClerkAPIResponseError.prototype);
        this.clerkError = true;
        this.message = message;
        this.status = status;
        this.errors = parseErrors(data);
    }
}
exports.ClerkAPIResponseError = ClerkAPIResponseError;
