// @clerk/backend-core/dist/cjs/api/endpoints/RedirectUrlApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedirectUrlAPI = void 0;
const path_1 = require("../../util/path");
const AbstractApi_1 = require("./AbstractApi");
const basePath = '/redirect_urls';
class RedirectUrlAPI extends AbstractApi_1.AbstractAPI {
    async getRedirectUrlList() {
        return this.APIClient.request({
            method: 'GET',
            path: basePath,
        });
    }
    async getRedirectUrl(redirectUrlId) {
        this.requireId(redirectUrlId);
        return this.APIClient.request({
            method: 'GET',
            path: (0, path_1.joinPaths)(basePath, redirectUrlId),
        });
    }
    async createRedirectUrl(params) {
        return this.APIClient.request({
            method: 'POST',
            path: basePath,
            bodyParams: params,
        });
    }
    async deleteRedirectUrl(redirectUrlId) {
        this.requireId(redirectUrlId);
        return this.APIClient.request({
            method: 'DELETE',
            path: (0, path_1.joinPaths)(basePath, redirectUrlId),
        });
    }
}
exports.RedirectUrlAPI = RedirectUrlAPI;
