// @clerk/backend-core/dist/cjs/api/endpoints/SessionApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionAPI = void 0;
const path_1 = require("../../util/path");
const AbstractApi_1 = require("./AbstractApi");
const basePath = '/sessions';
class SessionAPI extends AbstractApi_1.AbstractAPI {
    constructor() {
        super(...arguments);
        this.getSessionList = async (queryParams) => this.APIClient.request({
            method: 'GET',
            path: basePath,
            queryParams: queryParams,
        });
        this.getSession = async (sessionId) => {
            this.requireId(sessionId);
            return this.APIClient.request({
                method: 'GET',
                path: (0, path_1.joinPaths)(basePath, sessionId),
            });
        };
        this.revokeSession = async (sessionId) => {
            this.requireId(sessionId);
            return this.APIClient.request({
                method: 'POST',
                path: (0, path_1.joinPaths)(basePath, sessionId, 'revoke'),
            });
        };
        this.verifySession = async (sessionId, token) => {
            this.requireId(sessionId);
            return this.APIClient.request({
                method: 'POST',
                path: (0, path_1.joinPaths)(basePath, sessionId, 'verify'),
                bodyParams: { token },
            });
        };
        this.getToken = async (sessionId, template) => {
            this.requireId(sessionId);
            return (await this.APIClient.request({
                method: 'POST',
                path: (0, path_1.joinPaths)(basePath, sessionId, 'tokens', template || ''),
            })).jwt;
        };
    }
}
exports.SessionAPI = SessionAPI;
