// @clerk/backend-core/dist/mjs/api/ClerkBackendApi.js
import snakecaseKeys from 'snakecase-keys';
import { joinPaths } from '../util/path';
import { AllowlistIdentifierAPI, ClientAPI, EmailAPI, InvitationAPI, OrganizationAPI, RedirectUrlAPI, SessionAPI, SignInTokenAPI, SMSMessageAPI, UserAPI, } from './endpoints';
import { MISSING_API_CLIENT_ERROR, MISSING_API_KEY } from './errors';
import { deserialize } from './resources/Deserializer';
const DEFAULT_API_KEY = process.env.CLERK_API_KEY || '';
const DEFAULT_API_URL = process.env.CLERK_API_URL || 'https://api.clerk.dev';
const DEFAULT_API_VERSION = process.env.CLERK_API_VERSION || 'v1';
const INTERSTITIAL_URL = `${DEFAULT_API_URL}/${DEFAULT_API_VERSION}/internal/interstitial`;
export class ClerkBackendAPI {
    constructor(options) {
        this.options = options;
        const { apiClient, apiKey = DEFAULT_API_KEY, apiUrl = DEFAULT_API_URL, apiVersion = DEFAULT_API_VERSION, libName, libVersion, } = options;
        if (!apiClient) {
            throw Error(MISSING_API_CLIENT_ERROR);
        }
        this.apiClient = apiClient;
        this.apiKey = apiKey;
        this.apiUrl = apiUrl;
        this.apiVersion = apiVersion;
        this.userAgent = `${libName}@${libVersion}`;
    }
    async request(requestOptions) {
        if (!this.apiKey) {
            throw Error(MISSING_API_KEY);
        }
        const { path, method } = requestOptions;
        const url = joinPaths(this.apiUrl, this.apiVersion, path);
        const newRequestOptions = { method, url };
        newRequestOptions.queryParams = snakecaseKeys({
            ...this.getDefaultQueryParams(),
            ...requestOptions.queryParams,
        });
        newRequestOptions.headerParams = {
            ...this.getDefaultHeadersParams(),
            ...requestOptions.headerParams,
        };
        newRequestOptions.bodyParams = snakecaseKeys({
            ...this.getDefaultBodyParams(),
            ...requestOptions.bodyParams,
        }, { deep: false });
        const data = await this.apiClient.request(newRequestOptions);
        return deserialize(data);
    }
    fetchInterstitial() {
        return this.apiClient.request({
            url: INTERSTITIAL_URL,
            method: 'GET',
            headerParams: {
                ...this.getDefaultHeadersParams(),
                'Content-Type': 'text/html',
            },
        });
    }
    getDefaultQueryParams() {
        return {};
    }
    getDefaultHeadersParams() {
        return {
            Authorization: `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json',
            'User-Agent': this.userAgent,
            'X-Clerk-SDK': this.userAgent,
        };
    }
    getDefaultBodyParams() {
        return {};
    }
    get allowlistIdentifiers() {
        if (!this._allowlistIdentifierAPI) {
            this._allowlistIdentifierAPI = new AllowlistIdentifierAPI(this);
        }
        return this._allowlistIdentifierAPI;
    }
    get clients() {
        if (!this._clientAPI) {
            this._clientAPI = new ClientAPI(this);
        }
        return this._clientAPI;
    }
    get emails() {
        if (!this._emailAPI) {
            this._emailAPI = new EmailAPI(this);
        }
        return this._emailAPI;
    }
    get invitations() {
        if (!this._invitationAPI) {
            this._invitationAPI = new InvitationAPI(this);
        }
        return this._invitationAPI;
    }
    get organizations() {
        if (!this._organizationAPI) {
            this._organizationAPI = new OrganizationAPI(this);
        }
        return this._organizationAPI;
    }
    get redirectUrls() {
        if (!this._redirectUrlAPI) {
            this._redirectUrlAPI = new RedirectUrlAPI(this);
        }
        return this._redirectUrlAPI;
    }
    get sessions() {
        if (!this._sessionAPI) {
            this._sessionAPI = new SessionAPI(this);
        }
        return this._sessionAPI;
    }
    get signInTokens() {
        if (!this._signInTokenAPI) {
            this._signInTokenAPI = new SignInTokenAPI(this);
        }
        return this._signInTokenAPI;
    }
    get smsMessages() {
        if (!this._smsMessageAPI) {
            this._smsMessageAPI = new SMSMessageAPI(this);
        }
        return this._smsMessageAPI;
    }
    get users() {
        if (!this._userAPI) {
            this._userAPI = new UserAPI(this);
        }
        return this._userAPI;
    }
}
